src/deno_dir.rs

```diff
@@ -0,0 +1,398 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+use fs;
+use sha1;
+use std;
+use std::error::Error;
+use std::fs::File;
+use std::io::Write;
+use std::path::Path;
+use std::path::PathBuf;
+use std::result::Result;
+use url;
+use url::Url;
+
+#[cfg(test)]
+use tempfile::TempDir;
+
+pub struct DenoDir {
+  // Example: /Users/rld/.deno/
+  pub root: PathBuf,
+  // In the Go code this was called SrcDir.
+  // This is where we cache http resources. Example:
+  // /Users/rld/.deno/deps/github.com/ry/blah.js
+  pub gen: PathBuf,
+  // In the Go code this was called CacheDir.
+  // This is where we cache compilation outputs. Example:
+  // /Users/rld/.deno/gen/f39a473452321cacd7c346a870efb0e3e1264b43.js
+  pub deps: PathBuf,
+}
+
+impl DenoDir {
+  // Must be called before using any function from this module.
+  // https://github.com/ry/deno/blob/golang/deno_dir.go#L99-L111
+  pub fn new(custom_root: Option<&Path>) -> std::io::Result<DenoDir> {
+    // Only setup once.
+    let home_dir = std::env::home_dir().expect("Could not get home directory.");
+    let default = home_dir.join(".deno");
+
+    let root: PathBuf = match custom_root {
+      None => default,
+      Some(path) => path.to_path_buf(),
+    };
+    let gen = root.as_path().join("gen");
+    let deps = root.as_path().join("deps");
+
+    let deno_dir = DenoDir { root, gen, deps };
+    fs::mkdir(deno_dir.gen.as_ref())?;
+    fs::mkdir(deno_dir.deps.as_ref())?;
+
+    debug!("root {}", deno_dir.root.display());
+    debug!("gen {}", deno_dir.gen.display());
+    debug!("deps {}", deno_dir.deps.display());
+
+    Ok(deno_dir)
+  }
+
+  // https://github.com/ry/deno/blob/golang/deno_dir.go#L32-L35
+  pub fn cache_path(
+    self: &DenoDir,
+    filename: &str,
+    source_code: &str,
+  ) -> PathBuf {
+    let cache_key = source_code_hash(filename, source_code);
+    self.gen.join(cache_key + ".js")
+  }
+
+  fn load_cache(
+    self: &DenoDir,
+    filename: &str,
+    source_code: &str,
+  ) -> std::io::Result<String> {
+    let path = self.cache_path(filename, source_code);
+    debug!("load_cache {}", path.display());
+    fs::read_file_sync(&path)
+  }
+
+  pub fn code_cache(
+    self: &DenoDir,
+    filename: &str,
+    source_code: &str,
+    output_code: &str,
+  ) -> std::io::Result<()> {
+    let cache_path = self.cache_path(filename, source_code);
+    if cache_path.exists() {
+      Ok(())
+    } else {
+      let mut file = File::create(cache_path)?;
```

Comment:
This is a race condition w.r.t to exists() -- probably should create the file in exclusive mode.
I also worry what might happen is there are two processes and one reads the cache file while the other is in the midst of writing it. 
Not suggesting you fix it as part of this PR, but it'll have to be done some time.

Reply:
I added a todo comment.